fix(GridIntersect): fix DeprecationWarnings for Shapely2.0

dbrakenhoff opened the pull request:
- keep separate lists of cellids and geometries
- update intersect_* methods to use new STRtree api
- use .geoms for iteration over multipart geoms
- add deprecation warnings to old methods to generate list of geoms
- update tests
- add replacement function for deprecated descartes PolygonPatch

mwtoews commented:
I have not had a chance for a full review, but it's nice to see many of the warnings resolved. One new ShapelyDeprecationWarning introduced in the past 24-hours remains:
```
test_gridintersect.py: 46 warnings
  /home/runner/work/flopy/flopy/flopy/utils/gridintersect.py:178: ShapelyDeprecationWarning: STRtree will be changed in 2.0.0 and will not be compatible with versions < 2. Specifically: the items keyword argument will be removed.
    self.strtree = strtree.STRtree(self.geoms, self.cellids)
```
Also, I should mention that pre-releases for Shapely 2.0 are available, and can be installed with:

    pip install --upgrade --pre shapely

Some of the strategies in the algorithms in the flopy modules could be further improved. For instance, rather than using lists of geometries, use numpy arrays of geometries, since operations like `.intersects` are vectorized for 4 to 100x performance. For instance:
```python
# before shapely 2.0
list_result = [shp.intersects(geom) for geom in geoms]
# starting with shapely 2.0
array_result = shapely.intersects(shp, geoms)
```
arrays of pre-shapely 2.0 geometries can be constructed, but needs a [clever](https://github.com/geopandas/geopandas/blob/0ca19e160f1aad0f1538ecdb5c604166aa1adfdd/geopandas/_compat.py#L137-L169) [way](https://github.com/geopandas/geopandas/blob/0ca19e160f1aad0f1538ecdb5c604166aa1adfdd/geopandas/_vectorized.py#L109-L110) to avoid warnings.

dbrakenhoff commented:
Thanks for taking a look @mwtoews! I'll fix the remaining warning. 

My plan was to take a good look at the strategies and code when shapely 2.0 was released to figure out further performance improvements. I've written and tested a pygeos based GridIntersect class before, which easily outperformed the current implementation, so once the shapely 2.0 release becomes official it would be great to take advantage of that.

dbrakenhoff commented:
Alright, I made some changes to filter the Shapely2 warning. 

Couple of questions:
- I added a check whether the cellids are in order. Is this check necessary, or can i assume that they are always in order given a modflowgrid and how i convert the grid cells to geometries?
- @mwtoews could you take a look if I've modified the shapely warning filtering section at the top correctly? Since we require shapely>=1.8 now, I added an exception when that is not the case, and I removed one of the if-clauses related to shapely<1.8. I added a context manager for the shapely2 warnings.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -288,57 +311,67 @@ def intersect(
-                p.name = (i, j)
-                yield p
+                geoms.append(shapely_geo.Polygon(xy))
+                cellids.append(i * self.mfgrid.ncol + j)
```
this could be changed to 
```python
cellids.append(self.mfgrid.get_node([0, i, j])[0])
```
to avoid duplicate routines

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -385,18 +417,12 @@ def _rect_grid_to_shape_list(self):
+        return self._rect_grid_to_geoms_cellids()[0]
 
     def _vtx_grid_to_shape_list(self):
         """internal method, list of shapely polygons for vertex grids.
```
Reccomend adding a note in the docstrings that indicates that this routine was Deprecated in version 3.3.6.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -452,33 +484,43 @@ def filter_query_result(qresult, shp):
-    def sort_gridshapes(shape_iter):
-        """Sort query result by node id.
+    def sort_gridshapes(geoms, cellids):
+        """Sort geometries (from i.e. query result) by cell id.
```
Recommend adding a note in docstrings that this was Deprecated in version 3.3.6

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -557,7 +599,12 @@ def _intersect_point_shapely(
-                cellids.append(name)
+                # if structured calculated (i, j) cell address
+                if self.mfgrid.grid_type == "structured":
+                    i = cid // self.mfgrid.ncol
+                    j = cid - i * self.mfgrid.ncol
+                    cid = (i, j)
```
Recommend changing this to:
```python
cid = self.mfgrid.get_lrc([cid])[1:]
```
to avoid code duplication.

dbrakenhoff replied:
Thanks! Replaced the code with your suggestion.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -639,7 +691,12 @@ def _intersect_linestring_shapely(
                 vertices.append(verts)
-                cellids.append(name)
+                # if structured calculate (i, j) cell address
+                if self.mfgrid.grid_type == "structured":
+                    i = cid // self.mfgrid.ncol
+                    j = cid - i * self.mfgrid.ncol
+                    cid = (i, j)
```
Recommend changing this to:
```python
cid = self.mfgrid.get_lrc([cid])[1:]
```
to avoid code duplication.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -729,7 +791,12 @@ def _intersect_polygon_shapely(
                 vertices.append(verts)
-                cellids.append(name)
+                # if structured calculate (i, j) cell address
+                if self.mfgrid.grid_type == "structured":
+                    i = cid // self.mfgrid.ncol
+                    j = cid - i * self.mfgrid.ncol
+                    cid = (i, j)
```
Recommend changing this to:
```python
cid = self.mfgrid.get_lrc([cid])[1:]
```
to avoid code duplication.

jlarsen-usgs reviewed (changes requested):
@dbrakenhoff  Overall the pull request looks good. However, there are a couple little changes that I'd recommend before finalizing this PR.

1.) For newly deprecated methods, please add a note in the docstrings indicating that the method was deprecated in version 3.3.6. 

2.) Conversions between row, column to node and vice versa can be accomplished using existing built in functions on the StructuredGrid object. I've added examples of how to call these functions in the review comments.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -288,57 +311,67 @@ def intersect(
-                p.name = (i, j)
-                yield p
+                geoms.append(shapely_geo.Polygon(xy))
+                cellids.append(i * self.mfgrid.ncol + j)
```
this could be changed to 
```python
cellids.append(self.mfgrid.get_node([0, i, j])[0])
```
to avoid duplicate routines

dbrakenhoff replied:
done.

jlarsen-usgs commented on flopy/utils/gridintersect.py:
```diff
@@ -385,18 +417,12 @@ def _rect_grid_to_shape_list(self):
+        return self._rect_grid_to_geoms_cellids()[0]
 
     def _vtx_grid_to_shape_list(self):
         """internal method, list of shapely polygons for vertex grids.
```
Reccomend adding a note in the docstrings that indicates that this routine was Deprecated in version 3.3.6.

dbrakenhoff replied:
added message

mwtoews reviewed (approved):
I have a few suggestions, but I'll put them in a follow-up PR to review. Otherwise this looks great!
